src/ranged.cpp

```diff
@@ -2041,685 +2066,971 @@ std::vector<tripoint> target_handler::target_ui( spell &casting, const bool no_f
+    }
 
-        } else if( ( action == "PREV_TARGET" ) && ( target != -1 ) ) {
-            int newtarget = find_target( t, dst ) - 1;
-            if( newtarget < 0 ) {
-                newtarget = t.size() - 1;
-            }
-            dst = t[newtarget]->pos();
-        } else if( ( action == "NEXT_TARGET" ) && ( target != -1 ) ) {
-            int newtarget = find_target( t, dst ) + 1;
-            if( newtarget == static_cast<int>( t.size() ) ) {
-                newtarget = 0;
-            }
-            dst = t[newtarget]->pos();
-        } else if( action == "FIRE" ) {
-            if( casting.damage() > 0 && !confirm_non_enemy_target( dst ) ) {
-                continue;
-            }
-            find_target( t, dst );
-            break;
-        } else if( action == "CENTER" ) {
-            dst = src;
-            set_last_target( dst );
-            ret.clear();
-        } else if( action == "TOGGLE_SNAP_TO_TARGET" ) {
-            snap_to_target = !snap_to_target;
-        } else if( action == "QUIT" ) { // return empty vector (cancel)
-            ret.clear();
+    // Check if we were aiming at a tile or a (now missing) creature in a tile
+    // and still can aim at that tile.
+    cata::optional<tripoint> local_last_tgt_pos = cata::nullopt;
+    if( pc.last_target_pos ) {
+        tripoint local = g->m.getlocal( *pc.last_target_pos );
+        if( dist_fn( local ) > range ) {
+            // No luck
             pc.last_target_pos = cata::nullopt;
-            break;
-        }
-
-        // Make player's sprite flip to face the current target
-        if( dst.x > src.x ) {
-            g->u.facing = FD_RIGHT;
-        } else if( dst.x < src.x ) {
-            g->u.facing = FD_LEFT;
+        } else if( reentered ) {
+            local_last_tgt_pos = local;
         }
-
-    } while( true );
-
-    pc.view_offset = old_offset;
-
-    if( ret.empty() || ret.back() == pc.pos() ) {
-        return ret;
+    }
+    if( mode == TargetMode::Fire && pc.recoil == MAX_RECOIL ) {
+        // We've either moved away, used a bow or a gun with MASSIVE recoil. It doesn't really matter
+        // where we were aiming at, might as well start from scratch.
+        pc.last_target_pos = cata::nullopt;
+        local_last_tgt_pos = cata::nullopt;
+    }
+    if( local_last_tgt_pos ) {
+        new_dst = *local_last_tgt_pos;
+        return false;
     }
 
-    set_last_target( ret.back() );
+    // Try to find at least something
+    if( targets.empty() ) {
+        // The closest practice target
+        const std::vector<tripoint> nearby = closest_tripoints_first( src, range );
+        const auto target_spot = std::find_if( nearby.begin(), nearby.end(),
+        [&pc]( const tripoint & pt ) {
+            return g->m.tr_at( pt ).id == tr_practice_target && pc.sees( pt );
+        } );
 
-    const auto lt_ptr = pc.last_target.lock();
-    if( npc *const guy = dynamic_cast<npc *>( lt_ptr.get() ) ) {
-        if( casting.damage() > 0 ) {
```

Comment:
This condition being removed means that NPCs can be made angry when hit by healing spells.